### PR TITLE
add multi channel retrieval.

### PR DESCRIPTION
**Pull request by alalulu8668**
add multi channel to implement multiple retrieval for the query

### CODE REVIEW

File: bioimageio_chatbot/chatbot.py [hunk 1]
```diff
@@ -127,10 +129,10 @@ def create_customer_service(db_path):
     resource_item_stats = f"""Each item contains the following fields: {list(resource_items[0].keys())}\nThe available resource types are: {types}\nSome example tags: {tags}\nHere is an example: {resource_items[0]}"""
     class DocumentRetrievalInput(BaseModel):
         """Input for finding relevant documents from databases."""
-        query: str = Field(description="Query used to retrieve related documents.")
+        query: str = Field(description="Query to be used for retrieving relevant documents at different channel.")
```
Comment (oeway) [inline bioimageio_chatbot/chatbot.py:132]: This should also be a list, no?

Comment (oeway) [reply bioimageio_chatbot/chatbot.py:132]: Maybe create a list of ChannelQuery objects with channel_id and query as property

Comment (alalulu8668) [reply bioimageio_chatbot/chatbot.py]: > Maybe create a list of ChannelQuery objects with channel_id and query as property

Add ChannelQuery class, and integrate it as a list in DocumentRetrievalInput

File: bioimageio_chatbot/chatbot.py [hunk 2]
```diff
@@ -67,25 +67,27 @@ class ModelZooInfoScriptResults(BaseModel):
...
+    
 class DocumentSearchInput(BaseModel):
     """Results of document retrieval from documentation."""
     user_question: str = Field(description="The user's original question.")
     relevant_context: List[DocWithScore] = Field(description="Context chunks from the documentation")
-    user_info: str = Field(description="User info for personalize response.")
-    base_url: Optional[str] = Field(None, description="The base url of the documentation, used for resolve relative URL in the document and produce markdown links.")
-    format: Optional[str] = Field(None, description="The format of the document.")
+    user_info: Optional[str] = Field("", description="User info for personalize response.")
+    # base_url: Optional[str] = Field(None, description="The base url of the documentation, used for resolve relative URL in the document and produce markdown links.")
```
Comment (oeway) [inline bioimageio_chatbot/chatbot.py:89]: Remove here

File: bioimageio_chatbot/chatbot.py [hunk 3]
```diff
@@ -125,21 +127,27 @@ def create_customer_service(db_path):
...
+    class ChannelQuery(BaseModel):
+        """A query to a specific channel."""
+        query: str = Field(description="Query to be used for retrieving relevant documents at specific channel.")
+        channel_id: str = Field(description=f"Channel id used for retrieving. It MUST be the same as the user provided channel_id, and if not specified select one based on the query automatically. The available channels are:\n{channels_info}")
+        k: int = Field(description="The top-k number of documents to be retrieved from the channel, higher k number means more documents to retriev. Default 2, maximum 20.")
+    
     class DocumentRetrievalInput(BaseModel):
         """Input for finding relevant documents from databases."""
-        query: str = Field(description="Query used to retrieve related documents.")
+        query_channel: List[ChannelQuery] = Field(description="A list of queries to be used for retrieving relevant documents in one or more channels.")
```
Comment (oeway) [inline bioimageio_chatbot/chatbot.py:138]: How about queries?

File: bioimageio_chatbot/chatbot.py [hunk 4]
```diff
@@ -125,21 +127,27 @@ def create_customer_service(db_path):
...
         request: str = Field(description="User's request in details")
-        user_info: str = Field(description="Brief user info summary for personalized response, including name, background etc.")
-        channel_id: str = Field(description=f"It MUST be the same as the user provided channel_id, and if not specified select one automatically. The available channels are:\n{channels_info}")
+        user_info: Optional[str] = Field("",description="Brief user info summary for personalized response, including name, background etc.")
+        # channel_id: List[str] = Field(description=f"List of channels used for retrieving. It MUST be the same as the user provided channel_id, and if not specified select one or several based on the query automatically. The available channels are:\n{channels_info}")
```
Comment (oeway) [inline bioimageio_chatbot/chatbot.py:141]: Remove

File: bioimageio_chatbot/chatbot.py [hunk 5]
```diff
@@ -159,13 +167,21 @@ async def respond_to_user(question_with_history: QuestionWithHistory = None, rol
...
-            collection_info = collection_info_dict[req.channel_id]
-            print(f"Retrieving documents from database {req.channel_id} with query: {req.query}")
-            results_with_scores = await docs_store.asimilarity_search_with_relevance_scores(req.query, k=3)
-            docs_with_score = [DocWithScore(doc=doc.page_content, score=score, metadata=doc.metadata) for doc, score in results_with_scores]
-            print(f"Retrieved documents:\n{docs_with_score[0].doc[:20] + '...'} (score: {docs_with_score[0].score})\n{docs_with_score[1].doc[:20] + '...'} (score: {docs_with_score[1].score})\n{docs_with_score[2].doc[:20] + '...'} (score: {docs_with_score[2].score})")
-            search_input = DocumentSearchInput(user_question=req.request, relevant_context=docs_with_score, user_info=req.user_info, base_url=collection_info.get('base_url'), format=collection_info.get('format'))
+            # if len(req.query_channel) == 1:
+            #     retrieval_k = 3
+            # else:
+            #     retrieval_k = 2
```
Comment (oeway) [inline bioimageio_chatbot/chatbot.py:173]: Remove

**Pull request closed without merging**